### PR TITLE
Add gorelease Make target

### PR DESCRIPTION
**Pull request by pellared**
Fixes https://github.com/open-telemetry/opentelemetry-go/issues/4398

Semi-automated step that could help us avoiding common problems before releasing a new version.

While the tool is not perfect (see https://github.com/golang/go/issues/26420#issuecomment-1672883888), yet I still see some value.

E.g. 

```
gorelease in ./sdk/metric:
# go.opentelemetry.io/otel/sdk/metric
## incompatible changes
NewManualReader: changed from func(...ManualReaderOption) Reader to func(...ManualReaderOption) *ManualReader
NewPeriodicReader: changed from func(Exporter, ...PeriodicReaderOption) Reader to func(Exporter, ...PeriodicReaderOption) *PeriodicReader
Stream.AttributeFilter: removed
## compatible changes
ErrInstrumentName: added
InstrumentKind.String: added
ManualReader: added
PeriodicReader: added
Stream.AllowAttributeKeys: added
```

The `cannot find module providing package go.opentelemetry.io/otel/semconv/v1.21.0` is the most annoying issue as it kills the analysis for a lot of modules.

The workaround for it would be to postpone the `semconv` package creation just after the release and making a new release with just a `semconv` bump.

The good thing is that in https://github.com/open-telemetry/opentelemetry-go-contrib we should not have so many problems as we mostly use `replace` in `test` modules and `autoprop`. See https://github.com/open-telemetry/opentelemetry-go-contrib/pull/4169

Full output:

```
$make gorelease
gorelease in .:
gorelease: preparing to load packages for go.opentelemetry.io/otel: looking for missing dependencies: gorelease-load-module imports
        go.opentelemetry.io/otel/semconv/v1.21.0: cannot find module providing package go.opentelemetry.io/otel/semconv/v1.21.0

gorelease in ./bridge/opencensus:
# diagnostics
go.sum: one or more sums are missing. Run 'go mod tidy' to add missing sums.

# summary
Inferred base version: v0.39.0
Suggested version: v0.40.0 (with tag bridge/opencensus/v0.40.0)

gorelease in ./bridge/opencensus/test:
# diagnostics
go.sum: one or more sums are missing. Run 'go mod tidy' to add missing sums.

# summary
Inferred base version: v0.39.0
Suggested version: v0.40.0 (with tag bridge/opencensus/test/v0.40.0)

gorelease in ./bridge/opentracing:
# diagnostics
go.sum: one or more sums are missing. Run 'go mod tidy' to add missing sums.

# summary
Inferred base version: v1.16.0
Suggested version: v1.16.1 (with tag bridge/opentracing/v1.16.1)

gorelease in ./bridge/opentracing/test:
# diagnostics
go.sum: one or more sums are missing. Run 'go mod tidy' to add missing sums.

# summary
Inferred base version: v1.16.0
Suggested version: v1.17.0 (with tag bridge/opentracing/test/v1.17.0)

gorelease in ./example/fib:
gorelease: preparing to load packages for go.opentelemetry.io/otel/example/fib: looking for missing dependencies: gorelease-load-module imports
        go.opentelemetry.io/otel/example/fib imports
        go.opentelemetry.io/otel/semconv/v1.21.0: cannot find module providing package go.opentelemetry.io/otel/semconv/v1.21.0

gorelease in ./example/jaeger:
gorelease: preparing to load packages for go.opentelemetry.io/otel/example/jaeger: looking for missing dependencies: gorelease-load-module imports
        go.opentelemetry.io/otel/example/jaeger imports
        go.opentelemetry.io/otel/semconv/v1.21.0: cannot find module providing package go.opentelemetry.io/otel/semconv/v1.21.0

gorelease in ./example/namedtracer:
# diagnostics
go.sum: one or more sums are missing. Run 'go mod tidy' to add missing sums.

# summary
Inferred base version: v1.16.0
Suggested version: v1.17.0 (with tag example/namedtracer/v1.17.0)

gorelease in ./example/opencensus:
# diagnostics
go.sum: one or more sums are missing. Run 'go mod tidy' to add missing sums.

# summary
Inferred base version: v0.39.0
Suggested version: v0.40.0 (with tag example/opencensus/v0.40.0)

gorelease in ./example/otel-collector:
gorelease: preparing to load packages for go.opentelemetry.io/otel/example/otel-collector: looking for missing dependencies: gorelease-load-module imports
        go.opentelemetry.io/otel/example/otel-collector imports
        go.opentelemetry.io/otel/semconv/v1.21.0: cannot find module providing package go.opentelemetry.io/otel/semconv/v1.21.0
go: module github.com/golang/protobuf is deprecated: Use the "google.golang.org/protobuf" module instead.

gorelease in ./example/passthrough:
# diagnostics
go.sum: one or more sums are missing. Run 'go mod tidy' to add missing sums.

# summary
Inferred base version: v1.16.0
Suggested version: v1.17.0 (with tag example/passthrough/v1.17.0)

gorelease in ./example/prometheus:
# diagnostics
go.sum: one or more sums are missing. Run 'go mod tidy' to add missing sums.

# summary
Inferred base version: v0.39.0
Suggested version: v0.40.0 (with tag example/prometheus/v0.40.0)

gorelease in ./example/view:
# diagnostics
go.sum: one or more sums are missing. Run 'go mod tidy' to add missing sums.

# summary
Inferred base version: v0.39.0
Suggested version: v0.40.0 (with tag example/view/v0.40.0)

gorelease in ./example/zipkin:
gorelease: preparing to load packages for go.opentelemetry.io/otel/example/zipkin: looking for missing dependencies: gorelease-load-module imports
        go.opentelemetry.io/otel/example/zipkin imports
        go.opentelemetry.io/otel/semconv/v1.21.0: cannot find module providing package go.opentelemetry.io/otel/semconv/v1.21.0

gorelease in ./exporters/jaeger:
gorelease: preparing to load packages for go.opentelemetry.io/otel/exporters/jaeger: looking for missing dependencies: gorelease-load-module imports
        go.opentelemetry.io/otel/exporters/jaeger imports
        go.opentelemetry.io/otel/semconv/v1.21.0: cannot find module providing package go.opentelemetry.io/otel/semconv/v1.21.0

gorelease in ./exporters/otlp/internal/retry:
# summary
Inferred base version: v1.16.0
Suggested version: v1.16.1 (with tag exporters/otlp/internal/retry/v1.16.1)
gorelease in ./exporters/otlp/otlpmetric:
gorelease: go: loading module retractions for cloud.google.com/go/compute@v1.19.1: Get "https://proxy.golang.org/cloud.google.com/go/compute/@v/list": dial tcp 172.217.16.49:443: i/o timeout
go: loading module retractions for cloud.google.com/go/compute/metadata@v0.2.3: Get "https://proxy.golang.org/cloud.google.com/go/compute/metadata/@v/list": dial tcp 172.217.16.49:443: i/o timeout
go: loading module retractions for github.com/antihax/optional@v1.0.0: Get "https://proxy.golang.org/github.com/antihax/optional/@v/list": dial tcp 172.217.16.49:443: i/o timeout

gorelease in ./exporters/otlp/otlpmetric/otlpmetricgrpc:
# go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetricgrpc
## incompatible changes
New: changed from func(context.Context, ...Option) (go.opentelemetry.io/otel/sdk/metric.Exporter, error) to func(context.Context, ...Option) (*Exporter, error)
## compatible changes
Exporter: added

# diagnostics
go.sum: one or more sums are missing. Run 'go mod tidy' to add missing sums.

# summary
Inferred base version: v0.39.0
Suggested version: v0.40.0 (with tag exporters/otlp/otlpmetric/otlpmetricgrpc/v0.40.0)

gorelease in ./exporters/otlp/otlpmetric/otlpmetrichttp:
gorelease: listing versions of go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetrichttp: go: module go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetrichttp: Get "https://proxy.golang.org/go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetrichttp/@v/list": dial tcp 172.217.16.49:443: i/o timeout

gorelease in ./exporters/otlp/otlptrace:
# diagnostics
go.sum: one or more sums are missing. Run 'go mod tidy' to add missing sums.

# summary
Inferred base version: v1.16.0
Suggested version: v1.17.0 (with tag exporters/otlp/otlptrace/v1.17.0)

gorelease in ./exporters/otlp/otlptrace/otlptracegrpc:
# diagnostics
go.sum: one or more sums are missing. Run 'go mod tidy' to add missing sums.

# summary
Inferred base version: v1.16.0
Suggested version: v1.17.0 (with tag exporters/otlp/otlptrace/otlptracegrpc/v1.17.0)

gorelease in ./exporters/otlp/otlptrace/otlptracehttp:
# diagnostics
go.sum: one or more sums are missing. Run 'go mod tidy' to add missing sums.

# summary
Inferred base version: v1.16.0
Suggested version: v1.17.0 (with tag exporters/otlp/otlptrace/otlptracehttp/v1.17.0)

gorelease in ./exporters/prometheus:
# go.opentelemetry.io/otel/exporters/prometheus
## errors in release version:
exporter.go:60:32: undefined: metric.ManualReader

## compatible changes
(*Exporter).MarshalLog: added
WithoutCounterSuffixes: added

# diagnostics
go.sum: one or more sums are missing. Run 'go mod tidy' to add missing sums.

# summary
Inferred base version: v0.39.0
Cannot suggest a release version.
Errors were found.


gorelease in ./exporters/stdout/stdoutmetric:
# diagnostics
go.sum: one or more sums are missing. Run 'go mod tidy' to add missing sums.

# summary
Inferred base version: v0.39.0
Suggested version: v0.40.0 (with tag exporters/stdout/stdoutmetric/v0.40.0)

gorelease in ./exporters/stdout/stdouttrace:
# diagnostics
go.sum: one or more sums are missing. Run 'go mod tidy' to add missing sums.

# summary
Inferred base version: v1.16.0
Suggested version: v1.17.0 (with tag exporters/stdout/stdouttrace/v1.17.0)

gorelease in ./exporters/zipkin:
gorelease: preparing to load packages for go.opentelemetry.io/otel/exporters/zipkin: looking for missing dependencies: gorelease-load-module imports
        go.opentelemetry.io/otel/exporters/zipkin imports
        go.opentelemetry.io/otel/semconv/v1.21.0: cannot find module providing package go.opentelemetry.io/otel/semconv/v1.21.0

gorelease in ./metric:
# diagnostics
go.sum: one or more sums are missing. Run 'go mod tidy' to add missing sums.

# summary
Inferred base version: v1.16.0
Suggested version: v1.16.1 (with tag metric/v1.16.1)

gorelease in ./schema:
# summary
Inferred base version: v0.0.4
Suggested version: v0.1.0 (with tag schema/v0.1.0)
gorelease in ./sdk:
gorelease: preparing to load packages for go.opentelemetry.io/otel/sdk: looking for missing dependencies: gorelease-load-module imports
        go.opentelemetry.io/otel/sdk/resource imports
        go.opentelemetry.io/otel/semconv/v1.21.0: cannot find module providing package go.opentelemetry.io/otel/semconv/v1.21.0

gorelease in ./sdk/metric:
# go.opentelemetry.io/otel/sdk/metric
## incompatible changes
NewManualReader: changed from func(...ManualReaderOption) Reader to func(...ManualReaderOption) *ManualReader
NewPeriodicReader: changed from func(Exporter, ...PeriodicReaderOption) Reader to func(Exporter, ...PeriodicReaderOption) *PeriodicReader
Stream.AttributeFilter: removed
## compatible changes
ErrInstrumentName: added
InstrumentKind.String: added
ManualReader: added
PeriodicReader: added
Stream.AllowAttributeKeys: added

# go.opentelemetry.io/otel/sdk/metric/aggregation
## compatible changes
Base2ExponentialHistogram: added

# go.opentelemetry.io/otel/sdk/metric/metricdata
## compatible changes
ExponentialBucket: added
ExponentialHistogram: added
ExponentialHistogramDataPoint: added

# diagnostics
go.sum: one or more sums are missing. Run 'go mod tidy' to add missing sums.

# summary
Inferred base version: v0.39.0
Suggested version: v0.40.0 (with tag sdk/metric/v0.40.0)

gorelease in ./trace:
# diagnostics
go.sum: one or more sums are missing. Run 'go mod tidy' to add missing sums.

# summary
Inferred base version: v1.16.0
Suggested version: v1.16.1 (with tag trace/v1.16.1)
```

### CODE REVIEW

File: RELEASING.md [hunk 1]
```diff
@@ -21,6 +21,14 @@ make semconv-generate # Uses the exported TAG and OTEL_SEMCONV_REPO.
 This should create a new sub-package of [`semconv`](./semconv).
 Ensure things look correct before submitting a pull request to include the addition.
 
+## Breaking changes validation
+
+You can run `make gorelease` that runs
+[gorelease](https://pkg.go.dev/golang.org/x/exp/cmd/gorelease)
+to ensure that there are no unwanted changes done in the public API.
```
Comment (MrAlias) [inline RELEASING.md:28]: This should follow the line-break syntax of the doc:

```suggestion
You can run `make gorelease` that runs [gorelease](https://pkg.go.dev/golang.org/x/exp/cmd/gorelease) to ensure that there are no unwanted changes done in the public API.
```

File: RELEASING.md [hunk 2]
```diff
@@ -21,6 +21,14 @@ make semconv-generate # Uses the exported TAG and OTEL_SEMCONV_REPO.
...
+
+> You can check/report problems under <golang.org/issues/26420>.
```
Comment (MrAlias) [inline RELEASING.md:30]: Why is this quoted?

```suggestion
You can check/report problems with `gorelease` [here](golang.org/issues/26420).
```

Comment (pellared) [reply RELEASING.md:30]: No big reason 😉

**Review by MrAlias — APPROVED** [review]

**Comment by CharlieTLe** [comment]
Tried this out locally and got a different result:


```
make gorelease
cd ./internal/tools && \
        go build -o /Users/charliele/Work/repos/github/open-telemetry/opentelemetry-go/.tools/gorelease golang.org/x/exp/cmd/gorelease
no required module provides package golang.org/x/exp/cmd/gorelease; to add it:
        go get golang.org/x/exp/cmd/gorelease
make: *** [/Users/charliele/Work/repos/github/open-telemetry/opentelemetry-go/.tools/gorelease] Error 1
```

**Comment by CharlieTLe** [comment]
I needed to do

```
go work use internal/tools
```

since I was using go workspaces.

**Review by hanyuancheung — APPROVED** [review]

**Pull request merged**